Simple proxy for tavern replacement

WolframRavenwolf opened the pull request:
Continuing work on [Fixed presets](https://github.com/SillyTavern/SillyTavern/pull/932) and [[Feature Request] Add Simple Proxy functionality into Silly Tavern directly](https://github.com/SillyTavern/SillyTavern/issues/831):

- **Added Roleplay context template and updated instruct preset**
  - Updated Roleplay preset to work with last sequence and context templates.
- **Added simple-proxy-for-tavern context template and instruct preset**
  - The Roleplay preset is a simplified version of what simple-proxy-for-tavern does by default and tries to save tokens through some optimizations. It should achieve the same results in most cases without using up so many tokens.
  - Now, for testing purposes, here's an exact replica of simple-proxy-for-tavern's default verbose prompt template. We can decide later if we keep it (for proxy purists or potential special use cases) or drop it (to keep things simple) for the next public release.
- **Handle last_output_sequence like output_sequence**
  - Cleans up and sets stopping strings for last sequence analogously to output sequence.
- **Fix whitespace issues**
  - Extraneous newlines: There's code that adds the separatorSequence, which is a newline by default when "Wrap Sequences with Newline" is on and the Separator isn't set explicitly. And then there's code that joins the lines with newlines by default when "Wrap Sequences with Newline" is on. Both together cause double newlines, so I've appended the separator directly without joining, fixing this issue. Please double-check this part, though, to make sure it doesn't have unintended side effects.
  - ~~Space after name: When adding a name and a colon, we should always add a space, too! Otherwise the model has to add the space on its own, and that can cause problems since spaces are tokens, too, affected by repetition penalty and samplers. However, even though I added the space here, it seems to get trimmed somewhere else, so there's a follow-up change required.~~ ❎ Reverted: **no space after name**

**Additional Comments and Known Issues:**

- The original simple-proxy-for-tavern puts a linebreak at the very end of the whole prompt. SillyTavern doesn't do that, actually apparently it actively prevents it, trimming whitespace from the end of the prompt. Could we allow it for parity, in general or as an option? ✅ Done: **[End Prompt with Newline by StefanDanielSchwarz · Pull Request #969 · SillyTavern/SillyTavern](https://github.com/SillyTavern/SillyTavern/pull/969)**
- Now that we have multi-line input/output sequences, the cleanup and stopping string-adding code would probably work better if we added each line individually instead of as a multi-line string. ✅ Done: **deduplicated multi-line stopping strings**, **multi-line clean-up message**

WolframRavenwolf committed:
Added Roleplay context template and updated instruct preset

WolframRavenwolf committed:
Added simple-proxy-for-tavern context template and instruct preset

WolframRavenwolf committed:
Handle last_output_sequence like output_sequence

WolframRavenwolf committed:
Fix whitespace issues

WolframRavenwolf commented on public/scripts/power-user.js:
```diff
@@ -1070,8 +1070,8 @@ export function formatInstructModeChat(name, mes, isUser, isNarrator, forceAvata
-        : (power_user.instruct.wrap ? '\n' : '');
-    const textArray = includeNames ? [sequence, `${name}: ${mes}`, separatorSequence] : [sequence, mes, separatorSequence];
+        : separator;
+    const textArray = includeNames ? [sequence, `${name}: ${mes}` + separatorSequence] : [sequence, mes + separatorSequence];
```
**Extraneous newlines:** There's code that adds the separatorSequence, which is a newline by default when "Wrap Sequences with Newline" is on and the Separator isn't set explicitly. And then there's code that joins the lines with newlines by default when "Wrap Sequences with Newline" is on. Both together cause double newlines, so I've appended the separator directly without joining, fixing this issue. Please double-check this part, though, to make sure it doesn't have unintended side effects.

Cohee1207 replied:
Looks good so far.

WolframRavenwolf commented on public/scripts/power-user.js:
```diff
@@ -1097,7 +1097,7 @@ export function formatInstructModePrompt(name, isImpersonate, promptBias, name1,
 
     const separator = power_user.instruct.wrap ? '\n' : '';
-    let text = includeNames ? (separator + sequence + separator + `${name}:`) : (separator + sequence);
+    let text = includeNames ? (separator + sequence + separator + `${name}: `) : (separator + sequence);
```
When adding a name and a colon, we should always add a space, too! Otherwise the model has to add the space on its own, and that can cause problems since spaces are tokens, too, affected by repetition penalty and samplers. However, even though I added the space here, it seems to get trimmed somewhere else, so there's a follow-up change required.

Cohee1207 replied:
Okay, that makes sense. I think non-instruct prompt does that

Cohee1207 replied:
I gave it a second thought. Most tokens in sentencepiece start with a leading space, so by appending a space we greatly limit the number of tokens to be produced. Or there might be other side effects. So maybe it's better to remove it from here for now?

WolframRavenwolf replied:
Maybe it's a good thing if we limit the tokens to those which don't start with a leading space. Because those probably wouldn't be suitable tokens to start a new sentence with anyway.

But that's just my guess and would need further testing for confirmation. Alternatively, we could look at how Ooba and Kobold do it, and if they both do it the same way, it might be best to just do it that way as well.

Cohee1207 replied:
That I suppose would be our assignment for the next time.

WolframRavenwolf replied:
I've removed the space I added after the name because koboldcpp doesn't do that, either. More in another comment, but this conversation is thus resolved.

WolframRavenwolf committed:
deduplicated multi-line stopping strings

WolframRavenwolf commented:
**deduplicated multi-line stopping strings:** Now each line of multi-line input/output sequences is added individually as a stopping string, after eliminating duplicates.

WolframRavenwolf committed:
multi-line clean-up message

WolframRavenwolf commented:
**multi-line clean-up message:** Now each line of multi-line input/output sequences is cleaned up individually, which works better than trying to clean up multiple lines at once, as can be seen when using the simple-proxy-for-tavern instruct preset. I've only commented out the original code that did it for all lines at once, in case there might be a reason to use that at the same time, then we could just uncomment it to use both methods at once.

WolframRavenwolf committed:
fix default context

WolframRavenwolf commented:
**fix default context:** Eliminated empty lines when variables aren't set by putting the newlines inside the if clauses. (Doesn't look as pretty in the Story String text area, but fixes the actual prompt.) Also added "{{char}}'s personality: " back in, as was prefixed before, which is necessary for character cards that only list some comma-separated traits in here (see [Character Design | docs.ST.app](https://docs.sillytavern.app/usage/core-concepts/characterdesign/#personality-summary) which gives an example of just "Cheerful, cunning, provocative" for this part).

WolframRavenwolf commented:
@Cohee1207 That's it for today. 😎 Feel free to merge to Staging after reviewing the latest changes.

WolframRavenwolf committed:
no space after name

WolframRavenwolf commented:
I've removed the space I added after the name because koboldcpp doesn't do that, either. Didn't check ooba, but from a look at kobold's debugmode which shows tokenization, I agree that most tokens do start with a space and we'd not want to prevent those.

However, there might be a good case for inserting the space when Token Healing is in use. Here's an interesting article about it: [The Art of Prompt Design: Prompt Boundaries and Token Healing | by Scott Lundberg | Towards Data Science](https://towardsdatascience.com/the-art-of-prompt-design-prompt-boundaries-and-token-healing-3b2448b0be38)

With Token Healing, by inserting the space ourselves, we'd ensure the next token is one that starts with a space, which is exactly what we'd want. So might be worth to change it again at a later time once we can assume all backends to have that feature, but for the time being, let's stick to what's tried and true.
